### PR TITLE
[STORM-2687] Group Topology executors by network proximity needs and schedule them on network wise close slots

### PR DESCRIPTION
**Pull request by Ethanlm**
https://issues.apache.org/jira/browse/STORM-2687

This tries to schedule topology upstream and downstream executors closely.

Doing performance testing on this.

### CODE REVIEW

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java [hunk 1]
```diff
@@ -477,45 +414,136 @@ protected String nodeToRack(RAS_Node node) {
         List<ExecutorDetails> execsScheduled = new LinkedList<>();
 
         Map<String, Queue<ExecutorDetails>> compToExecsToSchedule = new HashMap<>();
-        for (Component component : componentMap.values()) {
-            compToExecsToSchedule.put(component.getId(), new LinkedList<ExecutorDetails>());
+        for (Map.Entry<String, Component> componentEntry: componentMap.entrySet()) {
```
Comment (revans2) [inline storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java]: Nit we only use the Component out of this and never the key, could we go back to just looping through the values like before?

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java [hunk 2]
```diff
@@ -477,45 +414,136 @@ protected String nodeToRack(RAS_Node node) {
...
+            Component component = componentEntry.getValue();
+            compToExecsToSchedule.put(component.getId(), new LinkedList<>());
             for (ExecutorDetails exec : component.getExecs()) {
                 if (unassignedExecutors.contains(exec)) {
                     compToExecsToSchedule.get(component.getId()).add(exec);
+                    LOG.info("{} has unscheduled executor {}", component.getId(), exec);
```
Comment (revans2) [inline storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java]: Could we remove this please?  Not sure it is needed anymore.

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java [hunk 3]
```diff
@@ -477,45 +414,136 @@ protected String nodeToRack(RAS_Node node) {
...
-            boolean flag = false;
-            do {
-                flag = false;
-                if (!currCompExesToSched.isEmpty()) {
-                    execsScheduled.add(currCompExesToSched.poll());
-                    flag = true;
-                }
+        }
+
+        LOG.info("The ordering result is {}", execsScheduled);
```
Comment (revans2) [inline storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java]: Could we remove this too?

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java [hunk 4]
```diff
@@ -477,45 +414,136 @@ protected String nodeToRack(RAS_Node node) {
...
+        return execsScheduled;
+    }
 
-                for (Component neighborComp : sortedNeighbors) {
-                    Queue<ExecutorDetails> neighborCompExesToSched =
-                        compToExecsToSchedule.get(neighborComp.getId());
-                    if (!neighborCompExesToSched.isEmpty()) {
-                        execsScheduled.add(neighborCompExesToSched.poll());
-                        flag = true;
+    private List<ExecutorDetails> takeExecutors(Component currComp, int numExecs,
```
Comment (revans2) [inline storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java]: Could you add some kind of javadoc to this explaining what it is trying to do?  It is not that obvious from just the code alone.

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java [hunk 5]
```diff
@@ -477,45 +414,136 @@ protected String nodeToRack(RAS_Node node) {
...
+                                                final Map<String, Component> componentMap,
+                                                final Map<String, Queue<ExecutorDetails>> compToExecsToSchedule) {
+        List<ExecutorDetails> execsScheduled = new ArrayList<>();
+        Queue<ExecutorDetails> currQueue = compToExecsToSchedule.get((currComp.getId()));
```
Comment (revans2) [inline storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java]: nit: there is an extra unneeded pair of '(' and ')'

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java [hunk 6]
```diff
@@ -477,45 +414,136 @@ protected String nodeToRack(RAS_Node node) {
...
+        Set<String> sortedChildren = getSortedChildren(currComp, componentMap);
+
+        execsScheduled.add(currQueue.poll());
```
Comment (revans2) [inline storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java]: Can `currQueue.poll()` ever return null?  How do we handle that if it does?

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java [hunk 7]
```diff
@@ -477,45 +414,136 @@ protected String nodeToRack(RAS_Node node) {
...
+                        return 1;
+                    } else {
+                        return -1;
+                    }
+                });
+        sortedChildren.addAll(children);
+        return sortedChildren;
+    }
+
+    private boolean isShuffleFromParentToChild(Component parent, Component child) {
```
Comment (revans2) [inline storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java]: Nit: could we rename this from `isShuffleFromParentToChild` to something more like `hasLocalityAwareGroupingFromParentToChild`?  I know it is longer, but in the future we may want to offer a way to expand this to more than just shuffle.

**Comment by HeartSaVioR** [comment]
@Ethanlm Do you have any updates?

**Comment by Ethanlm** [comment]
@HeartSaVioR  I got stuck on measuring the performance gain and didn't find a good way to do it. Currently occupied by something else. Will try to get back to it later

**Comment by Ethanlm** [comment]
Closing this in favor of https://github.com/apache/storm/pull/3265

**Pull request closed without merging**